### PR TITLE
Allow --stdin for nbconvert (alternative command line interface)

### PR DESCRIPTION
**Pull request by parleur**
Follow #166 and #164, @akhmerov @Carreau @minrk 
Here the command line interface is:

| nbconvert --stdin --to html 
write in "notebook.html" by default
| nbconvert --stdin --to html --stdout
write to stdout

The corner in this API is default basename I am forced to introduce. Public APIs are changed but not broken, just extended. Right now help, and docstring havn't been double cheked (coming ...).

### CODE REVIEW

**Comment by parleur** [comment]
Maybe random basename could be an option ...

**Comment by akhmerov** [comment]
This replaces #166, right?

**Comment by Carreau** [comment]
No I think it is an alternative where --stdin does not imply --stdout

And the API change where you can pass a buffer as a kwarg instead of the first arg eing (filename or buffer)

**Comment by parleur** [comment]
Sorry to be unclear ... It's a more more invasive alternative for #166

It responds to your comments about cleaning the code (introduce API extention `input_buffer` which is clearer than previously.

>While it's probably not very relevant to convert from stdin to a file, you're tying two unrelated options together — source of the notebook and output of the notebook. It might be cleaner to not do so (otoh it makes the interface more complex)

 And modify the command line interface as you pointed.

**Comment by parleur** [comment]
As @Carreau said it modify the API (API just extended). But could be balanced by making the code a bit more clearer than having `notebook_filename` being a buffer or a filename and testing external flags to kwow what we are dealing with.

File: nbconvert/nbconvertapp.py [hunk 1]
```diff
@@ -69,6 +69,13 @@ def validate(self, obj, value):
          "(the default behaviour is to abort conversion). This flag "
          "is only relevant if '--execute' was specified, too.")
         ),
+    'stdin' : (
+        {'NbConvertApp' : {
+            'from_stdin' : True,
+            }
+        },
+        "read a single notebook file from stdin, and write the converted result to stdout. Implies `--stdout`"
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: The description seems obsolete.

File: nbconvert/nbconvertapp.py [hunk 2]
```diff
@@ -323,19 +335,23 @@ def init_single_notebook_resources(self, notebook_filename):
...
 
         Exports the notebook to a particular format according to the specified
         exporter. This function returns the output and (possibly modified)
         resources from the exporter.
 
+        notebook_filename: a filename or buffer (for instance from stdin using '--stdin' option)
         """
         try:
-            output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
+            if input_buffer:
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: `if input_buffer is not None`
(same in analogous places)

File: nbconvert/nbconvertapp.py [hunk 3]
```diff
@@ -170,12 +177,13 @@ def _classes_default(self):
...
-                      help="""Instance of the writer class used to write the 
+    writer = Instance('nbconvert.writers.base.WriterBase',
+                      help="""Instance of the writer class used to write the
                       results of the conversion.""", allow_none=True)
-    writer_class = DottedObjectName('FilesWriter', config=True, 
-                                    help="""Writer class used to write the 
+    writer_class = DottedObjectName('FilesWriter', config=True,
+                                    help="""Writer class used to write the
                                     results of the conversion""")
+    from_stdin   = Bool(False, config=True, help="read a single notebook from stdin.")
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: This is located in a confusing place in the file: it's definitely not a writer-specific variable. I suggest to move to the group of `# Other configurable variables`.

File: nbconvert/nbconvertapp.py [hunk 4]
```diff
@@ -323,19 +335,23 @@ def init_single_notebook_resources(self, notebook_filename):
...
 
-    def export_single_notebook(self, notebook_filename, resources):
+    def export_single_notebook(self, notebook_filename, resources, input_buffer=None):
         """Step 2: Export the notebook
 
         Exports the notebook to a particular format according to the specified
         exporter. This function returns the output and (possibly modified)
         resources from the exporter.
 
+        notebook_filename: a filename or buffer (for instance from stdin using '--stdin' option)
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: Obsolete, cannot be a buffer.

File: nbconvert/nbconvertapp.py [hunk 5]
```diff
@@ -323,19 +335,23 @@ def init_single_notebook_resources(self, notebook_filename):
...
-            output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
+            if input_buffer is not None:
+                output, resources = self.exporter.from_file(input_buffer, resources=resources)
+            else:
+                output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
         except ConversionException:
-            self.log.error("Error while converting '%s'", notebook_filename, exc_info=True)
-            self.exit(1)
+            self.log.error("Error while converting notebook", exc_info=True)
+            self.exit('Could not convert notebook.')
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: This change seems unrelated to the PR; is it an improvement?

File: nbconvert/nbconvertapp.py [hunk 6]
```diff
@@ -369,18 +385,27 @@ def postprocess_single_notebook(self, write_results):
...
             4. (Maybe) postprocess the written file
-
+            
+            If input_buffer is not None, convertion is done using buffer as source ignoring
+            the notebook_filename argument.
         """
-        self.log.info("Converting notebook %s to %s", notebook_filename, self.export_format)
-        resources = self.init_single_notebook_resources(notebook_filename)
-        output, resources = self.export_single_notebook(notebook_filename, resources)
+        if not input_buffer:
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: `if input_buffer is None`

File: nbconvert/nbconvertapp.py [hunk 7]
```diff
@@ -369,18 +385,27 @@ def postprocess_single_notebook(self, write_results):
...
+            self.log.info("Converting notebook %s to %s", notebook_filename, self.export_format)
+            resources = self.init_single_notebook_resources(notebook_filename)
+            output, resources = self.export_single_notebook(notebook_filename, resources)
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: This line can be taken out of `if` statement (together with the matching line in `else`.

File: nbconvert/nbconvertapp.py [hunk 8]
```diff
@@ -369,18 +385,27 @@ def postprocess_single_notebook(self, write_results):
...
+        else:
+            notebook_filename = "notebook"
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: This seems unnecessary. You can assume it's a legitimate name even if `input_buffer` is set. Then you can just provide the value `notebook` in line 440.

File: nbconvert/nbconvertapp.py [hunk 9]
```diff
@@ -398,18 +423,21 @@ def convert_notebooks(self):
...
             sys.exit(-1)
 
         # convert each notebook
-        for notebook_filename in self.notebooks:
-            self.convert_single_notebook(notebook_filename)
+        if not self.from_stdin:
+            for notebook_filename in self.notebooks:
+                self.convert_single_notebook(notebook_filename)
+        else:
+            self.convert_single_notebook(None, input_buffer=sys.stdin)
```
Comment (akhmerov) [inline nbconvert/nbconvertapp.py]: Related to comment about l404: 

`self.convert_single_notebook('notebook', input_buffer=sys.stdin)`

**Comment by akhmerov** [comment]
Other things still to consider potentially:
* fixing the whitespace changes by `git checkout -i`.
* whatsnew entry.

Other than that and above comments, everything seems perfect to me.

**Comment by parleur** [comment]
Thanks a lot for all your comments. I am keep trying to suppress my coding glitches. 

What do mean by 'whatsnew entry' ? Precisely where it is supposed to be written ?

**Comment by akhmerov** [comment]
> What do mean by 'whatsnew entry' ? Precisely where it is supposed to be written ?

Just a description for changelog, @Carreau should know where it needs to go.

**Comment by Carreau** [comment]
> @Carreau should know where it needs to go.

`docs/source/changelog.rst` I guess. 

Thanks !

**Comment by akhmerov** [comment]
Happy new year everyone! Should this be merged? I don't see any further problems.

**Comment by Carreau** [comment]
let's get notebook 4.1, and then pass on all those things that have waited too long.

But in principle,+1.

File: nbconvert/nbconvertapp.py [hunk 10]
```diff
@@ -323,16 +330,21 @@ def init_single_notebook_resources(self, notebook_filename):
...
-    def export_single_notebook(self, notebook_filename, resources):
+    def export_single_notebook(self, notebook_filename, resources, input_buffer=None):
         """Step 2: Export the notebook
 
         Exports the notebook to a particular format according to the specified
         exporter. This function returns the output and (possibly modified)
         resources from the exporter.
 
+        notebook_filename: a filename
+        input_buffer: a containing notebook buffer, if not None notebook_filename is ignored
```
Comment (takluyver) [inline nbconvert/nbconvertapp.py]: This description wasn't completely clear to me. The convention I'm used to in Python is to call these 'file like objects'. Annoyingly, there's a lot of 'file like objects' with slightly different properties, so it's a good idea to clarify, e.g.

> a readable file like object returning unicode

File: nbconvert/nbconvertapp.py [hunk 11]
```diff
@@ -403,13 +421,17 @@ def convert_notebooks(self):
...
 
         # convert each notebook
-        for notebook_filename in self.notebooks:
-            self.convert_single_notebook(notebook_filename)
+        if not self.from_stdin:
+            for notebook_filename in self.notebooks:
+                self.convert_single_notebook(notebook_filename)
+        else:
+            # default name when conversion from stdin
+            self.convert_single_notebook("notebook.ipynb", input_buffer=sys.stdin)
```
Comment (takluyver) [inline nbconvert/nbconvertapp.py]: Unfortunately, I think this will run into Unicode problems on Python 3 in non-UTF-8 systems, primarily Windows. Python will use the system's default encoding to try to decode stdin, and if a notebook is piped in containing a non-ascii character, and it tries to decode it using cp1252 or something, it will blow up.

Have a look at `nbconvert.utils.io.unicode_std_stream` for what we need to do here. That's for output, but basically the same steps should work for stdin.

**Comment by parleur** [comment]
Happy new year, and sorry for the latency.

File: nbconvert/tests/test_nbconvertapp.py [hunk 12]
```diff
@@ -315,8 +316,33 @@ def test_fenced_code_blocks_markdown(self):
             output2, _ = self.nbconvert('--to markdown --stdout notebook_jl.ipynb')
             assert '```julia' in output2  # shouldn't have language
             assert "```" in output2  # but should also plain ``` to close cell
-
-        pass
+    
+    def test_convert_from_stdin_to_stdout(self):
+        """
+        Verify that conversion can be done via stdin, and that 
+        --stdin  implies --stdout.
```
Comment (takluyver) [inline nbconvert/tests/test_nbconvertapp.py]: This docstring is misleading now - `--stdin` no longer implies `--stdout`.

File: nbconvert/tests/test_nbconvertapp.py [hunk 13]
```diff
@@ -315,8 +316,33 @@ def test_fenced_code_blocks_markdown(self):
...
+            with io.open('notebook1.ipynb') as f:
+                notebook = f.read().encode()
+                output1, _ = self.nbconvert('--to markdown --stdin --stdout', stdin=notebook)
+            assert_not_in('```python', output1) # shouldn't have language
+            assert_in("```", output1) # but should have fenced blocks
+
+    def test_convert_from_stdin(self):
+        """
+        Verify that conversion can be done via stdin, and that 
+        --stdin  implies --stdout.
```
Comment (takluyver) [inline nbconvert/tests/test_nbconvertapp.py]: And again

File: nbconvert/utils/io.py [hunk 14]
```diff
@@ -31,3 +31,20 @@ def unicode_std_stream(stream='stdout'):
...
+
+    This ignores environment variables and default encodings, to reliably read unicode from stdin.
+
+    ::
+
+        totreat = unicode_stdin_stream().read()
+    """
+    stream  = sys.stdin
+    if PY3:
+        stream_b = stream.buffer
```
Comment (takluyver) [inline nbconvert/utils/io.py]: I think this should do the same try/except AttributeError that we do in `unicode_std_stream()` above. This is to deal with the case where someone has set `sys.stdin` to another object, like a `StringIO` instance. That's less common for stdin than for stdout, but it's still a valid thing to do.

**Comment by takluyver** [comment]
Apart from a few minor comments, I think this looks good. Hopefully we can finally get this merged soon!

**Comment by Carreau** [comment]
+1, looking for ward to get back on "dev and break all the things".

File: nbconvert/utils/io.py (hunk 14 above)
Comment (parleur) [inline nbconvert/utils/io.py]: I think it is done in tests ... that's why I dealt with in a proper way.

**Comment by Carreau** [comment]
+1, thanks for you patience. 

5.0 of 4.2 ? Is does not break API.

**Comment by takluyver** [comment]
I don't think we have any really major breaking changes planned for nbconvert, so I'm not sure how we want to handle the version numbering. Are 4.x releases roll-ups of bugfixes, as we've traditionally done things, and 5.0 will be a feature release, or do we want to do a longer series of 4.x feature releases and switch to 5.0 when we want to break things. The latter is more semver style, but I don't think semver should be the only way to do things.

**Comment by Carreau** [comment]
I would do 4.2 with a few new features, as we already have new things, like `--to Qualified.names`.

I just asked because the current version shows 4.2.dev, and you tagged 5.0. 

I'm find starting a 5.0.dev and having a 4.2 branch also. 

If you are fine with 4.2 with new features, I woudl just merge that now.

**Comment by minrk** [comment]
I think in general, across all repos, we don't bump major until we start merging breaking changes. And in general, we should be requiring stronger justifications for any breaking changes we make, since any breaking change may force a coordinated release. We only need to start backporting once we've started down the path of breaking changes, but want to keep releasing bugfixes for prior versions.

I would expect most repos other than notebook/ipython to spend much of their time in minor releases adding features and not breaking stuff.

**Comment by takluyver** [comment]
Great, then I think this can be landed. Thanks @parleur for your patience in working on this!
